C node restart

snazy opened the pull request:
Adds:
* Cassandra server shutdown
* Cassandra server start
* Cassandra server restart

To do that, a `TargetRunState` enum is introduced. It knows three states: `RUN`=node has to run, `STOP`=node should not run, `RESTART`=stop and start node (sets state to `RUN` afterwards)

Since a new task ID (`shutdown`) is being introduced (and a follow-up PR introduces a new state `TERMINATE` along with another new task `terminate`) I've change refactored the `metadataTask` and `serverTask` fields to a general-purpose list of `CassandraNode.tasks`.

snazy commented:
Note: the "replace node PR" depends on this PR - so I'll wait until this one's been merged.

BenWhitehead commented on cassandra-executor/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraExecutor.java:
```diff
@@ -124,19 +124,51 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo task) {
+                            } catch (Exception e) {
+                                LOGGER.error(taskIdMarker, "Failed to stop Cassandra daemon - forcing process destroy", e);
+                            }
+                        }
```
we probably want to add an `else` here that will fall back to trying to run `kill -9 $(cat cassandra.pid)` if we're not able to use the jmx `stopDaemon` method.

We should also remove `cassandra.pid` when we shutdown the server.

snazy replied:
Damn - have read the JDK source wrong (thought it already does a `kill -9`). Java8 has a new `Process.destroyForcibly()` method for that. But we should try to do a `destroy()` before `kill -9` if JMX doesn't work for us.

BenWhitehead replied:
Sure that sounds good to try destroy before kill -9.  I think they're just stages of trying to make sure it's shutdown.

BenWhitehead commented on cassandra-executor/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraExecutor.java:
```diff
@@ -343,6 +379,9 @@ private void safeShutdown() {
             process = null;
+
+            // send health check that process is no longer alive
+            healthCheck(driver);
```
What is the motivation for sending a healthcheck message when we're shutting down the server.  Isn't "shut down" and "unhealthy" two different things in the scenario that the scheduler asked that the process be shut down?

snazy replied:
Generally yes - but that way the scheduler immediately knows that the server is dead.
But you're right - I forgot some code to actually update the current health check when it receives that the task is no longer running.

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java:
```diff
@@ -489,4 +496,62 @@ private void writeClusterJobNode(JsonGenerator json, CassandraFrameworkProtos.No
+
+    // node run / stop / restart
+
+    @GET
```
Change to `@POST` all write operations should be `@POST` or `@PUT`

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java:
```diff
@@ -489,4 +496,62 @@ private void writeClusterJobNode(JsonGenerator json, CassandraFrameworkProtos.No
+        return nodeStatusUpdate(cassandraNode);
+    }
+
+    @GET
```
Change to `@POST` all write operations should be `@POST` or `@PUT`

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/ApiController.java:
```diff
@@ -489,4 +496,62 @@ private void writeClusterJobNode(JsonGenerator json, CassandraFrameworkProtos.No
+        return nodeStatusUpdate(cassandraNode);
+    }
+
+    @GET
```
Change to `@POST` all write operations should be `@POST` or `@PUT`

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java:
```diff
@@ -158,24 +158,27 @@ public void removeTask(@NotNull final String taskId, Protos.TaskStatus status) {
+            changed = true;
+            switch (nodeTask.getTaskType()) {
+                case METADATA:
+                    // TODO shouldn't we also assume that the server task is no longer running ??
```
Removing of executor info and taking the server with it should be handled by the status update from `executorLost` so that once `executorLost` is fixed in mesos it can still be handled there.

snazy replied:
Noted a TODO for later.

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java:
```diff
@@ -158,24 +158,27 @@ public void removeTask(@NotNull final String taskId, Protos.TaskStatus status) {
+                    removeExecutorMetadata(nodeTask.getExecutorId());
+                    builder.clearTasks();
+                    break;
+                case SERVER:
```
redundant switch branch, `changed` is set to `true` before the switch begins.

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java:
```diff
@@ -270,7 +272,7 @@ public void recordHealthCheck(@NotNull final String executorId, @NotNull final H
                 LOGGER.info(
                     "health check result unhealthy for node: {}. Message: '{}'",
-                    nodeOpt.get().getCassandraNodeExecutor().getExecutorId(),
+                    nodeOpt.get().hasCassandraNodeExecutor() ? nodeOpt.get().getCassandraNodeExecutor().getExecutorId() : "??",
```
A better fallback would be `nodeOpt.get().getHostname()` instead of `"??"`

BenWhitehead commented on cassandra-scheduler/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraCluster.java:
```diff
@@ -330,14 +332,29 @@ public TasksForOffer getTasksForOffer(@NotNull final Protos.Offer offer) {
+                            case RUN:
+                                break;
+                            case STOP:
+                                if (LOGGER.isDebugEnabled())
```
if withouth `{` `}`

BenWhitehead commented:
Mesos already provides a primitive for killing tasks.

From the scheduler you can call `driver.killTask(taskID)` the executor will then receive the `taskId` via `io.mesosphere.mesos.frameworks.cassandra.CassandraExecutor#killTask` where we can then kill the server.

I think the improved state tracking of the server is good, but we won't need the custom task type to shutdown the server. I think you can leverage this by moving the `case CASSANDRA_SERVER_SHUTDOWN` branch in the executor from `launchTask` to `killTask` and everything should be the same.

You will probably have to add a method to `CassandraCluster` to the effect of `getTaskIDsToKill(offer)` and then call that before calling `getTasksForOffer()` from `io.mesosphere.mesos.frameworks.cassandra.CassandraScheduler#evaluateOffer`

BenWhitehead commented:
several if without `{` `}` please add braces to all control structures.

snazy commented:
Okay - will exchange `CASSANDRA_SERVER_SHUTDOWN` with `killTask` (didn't knew that) - feels better. Will to the outstanding review changes tomorrow.

snazy commented:
Pushed rebased branch with review issues fixed.
Also changed how the C* daemon is started:
* removed `-p cassandra.pid` because it's not written when `-f` is given
* takes PID from `java.lang.UNIXProcess`
* also passes PID to scheduler
* honors `LaunchNodeException` and returns `TASK_ERROR`
* checks if process stays alive for at least 500ms and returns `TASK_ERROR` if not (might occur if some settings prevent C* to start)

BenWhitehead commented on cassandra-executor/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraExecutor.java:
```diff
@@ -223,6 +238,49 @@ public void killTask(final ExecutorDriver driver, final TaskID taskId) {
+        }
+    }
+
+    private void killCassandraDaemon(ExecutorDriver driver) {
```
Please move this method down with the other private methods.

BenWhitehead commented:
I'm not crazy about the UNIXProcess reflection stuff, but it's probably okay until we update to Java 1.8.

BenWhitehead commented:
One nit about code structure.  Then should be ready to merge after a rebase.

snazy commented:
Rebased
